// alipay-sdk/dist/commonjs/antcertutil.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.loadPublicKeyFromPath = loadPublicKeyFromPath;
exports.loadPublicKey = loadPublicKey;
exports.getSNFromPath = getSNFromPath;
exports.getSN = getSN;
const node_fs_1 = __importDefault(require("node:fs"));
const node_crypto_1 = require("node:crypto");
const bignumber_js_1 = require("bignumber.js");
const x509_1 = require("@fidm/x509");
/** 从公钥证书文件里读取支付宝公钥 */
function loadPublicKeyFromPath(filePath) {
    const fileData = node_fs_1.default.readFileSync(filePath);
    const certificate = x509_1.Certificate.fromPEM(fileData);
    return certificate.publicKeyRaw.toString('base64');
}
/** 从公钥证书内容或 Buffer 读取支付宝公钥 */
function loadPublicKey(content) {
    const pemContent = typeof content === 'string' ? Buffer.from(content) : content;
    const certificate = x509_1.Certificate.fromPEM(pemContent);
    return certificate.publicKeyRaw.toString('base64');
}
/** 从证书文件里读取序列号 */
function getSNFromPath(filePath, isRoot = false) {
    const fileData = node_fs_1.default.readFileSync(filePath);
    return getSN(fileData, isRoot);
}
/** 从上传的证书内容或 Buffer 读取序列号 */
function getSN(fileData, isRoot = false) {
    const pemData = typeof fileData === 'string' ? Buffer.from(fileData) : fileData;
    if (isRoot) {
        return getRootCertSN(pemData);
    }
    const certificate = x509_1.Certificate.fromPEM(pemData);
    return getCertSN(certificate);
}
/** 读取序列号 */
function getCertSN(certificate) {
    const { issuer, serialNumber } = certificate;
    const principalName = issuer.attributes
        .reduceRight((prev, curr) => {
        const { shortName, value } = curr;
        const result = `${prev}${shortName}=${value},`;
        return result;
    }, '')
        .slice(0, -1);
    const decimalNumber = new bignumber_js_1.BigNumber(serialNumber, 16).toString(10);
    const SN = (0, node_crypto_1.createHash)('md5')
        .update(principalName + decimalNumber, 'utf8')
        .digest('hex');
    return SN;
}
/** 读取根证书序列号 */
function getRootCertSN(rootContent) {
    const certificates = x509_1.Certificate.fromPEMs(rootContent);
    let rootCertSN = '';
    certificates.forEach(item => {
        if (item.signatureOID.startsWith('1.2.840.113549.1.1')) {
            const SN = getCertSN(item);
            if (rootCertSN.length === 0) {
                rootCertSN += SN;
            }
            else {
                rootCertSN += `_${SN}`;
            }
        }
    });
    return rootCertSN;
}
