// alipay-sdk/dist/commonjs/form.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AlipayFormData = void 0;
class AlipayFormData {
    method;
    files;
    fields;
    constructor() {
        this.fields = [];
        this.files = [];
        this.method = 'post';
    }
    getFields() { return this.fields; }
    getFiles() { return this.files; }
    getMethod() { return this.method; }
    /**
     * 设置 method
     * post、get 的区别在于 post 会返回 form 表单，get 返回 url
     */
    setMethod(method) {
        this.method = method.toLowerCase();
    }
    /**
     * 增加字段
     * @param fieldName 字段名
     * @param fieldValue 字段值
     */
    addField(fieldName, fieldValue) {
        if (isJSONString(fieldValue)) {
            // 当 fieldValue 为 json 字符串时，解析出 json
            this.fields.push({ name: fieldName, value: JSON.parse(fieldValue) });
        }
        else {
            this.fields.push({ name: fieldName, value: fieldValue });
        }
    }
    /**
     * 增加文件
     * @param fieldName 文件字段名
     * @param fileName 文件名
     * @param filePath 文件绝对路径，或者文件流，又或者文件内容 Buffer
     */
    addFile(fieldName, fileName, filePath) {
        const file = {
            fieldName,
            name: fileName,
        };
        if (typeof filePath === 'string') {
            file.path = filePath;
        }
        else if (Buffer.isBuffer(filePath)) {
            file.content = filePath;
        }
        else {
            file.stream = filePath;
        }
        this.files.push(file);
    }
}
exports.AlipayFormData = AlipayFormData;
// forked form https://github.com/joaquimserafim/is-json/blob/master/index.js#L6
function isJSONString(value) {
    if (typeof value !== 'string')
        return false;
    value = value.replace(/\s/g, '').replace(/\n|\r/, '');
    if (/^\{(.*?)\}$/.test(value)) {
        return /"(.*?)":(.*?)/g.test(value);
    }
    if (/^\[(.*?)\]$/.test(value)) {
        return value.replace(/^\[/, '')
            .replace(/\]$/, '')
            .replace(/},{/g, '}\n{')
            .split(/\n/)
            .map((s) => { return isJSONString(s); })
            .reduce(function (_prev, curr) { return !!curr; });
    }
    return false;
}
